pkg/asset/agent/manifests/nmstateconfig.go

```diff
@@ -242,41 +243,68 @@ func (n *NMStateConfig) validateNMStateLabels() field.ErrorList {
 	return allErrs
 }
 
-func getFirstIP(nmStateConfig *nmStateConfig) string {
+func getFirstIP(nmstateRaw []byte) (string, error) {
+	var nmStateConfig nmStateConfig
+	err := yaml.Unmarshal(nmstateRaw, &nmStateConfig)
+	if err != nil {
+		return "", fmt.Errorf("error unmarshalling NMStateConfig: %w", err)
+	}
+
 	for _, intf := range nmStateConfig.Interfaces {
 		for _, addr4 := range intf.IPV4.Address {
 			if addr4.IP != "" {
-				return addr4.IP
+				return addr4.IP, nil
 			}
 		}
 		for _, addr6 := range intf.IPV6.Address {
 			if addr6.IP != "" {
-				return addr6.IP
+				return addr6.IP, nil
 			}
 		}
 	}
-	return ""
+
+	return "", nil
 }
 
-// GetNodeZeroIP retrieves the first IP from the user provided NMStateConfigs to set as the node0 IP
-func GetNodeZeroIP(nmStateConfigs []*aiv1beta1.NMStateConfig) (string, error) {
-	for i := range nmStateConfigs {
-		var nmStateConfig nmStateConfig
-		err := yaml.Unmarshal(nmStateConfigs[i].Spec.NetConfig.Raw, &nmStateConfig)
+// GetNodeZeroIP retrieves the first IP to be set as the node0 IP.
+// The method prioritizes the search by trying to scan first the NMState configs defined
+// in the agent-config hosts - so that it would be possible to skip the worker nodes - and then
+// the NMStateConfig.
+func GetNodeZeroIP(agentConfig *agenttypes.Config, nmStateConfigs []*aiv1beta1.NMStateConfig) (string, error) {
+	rawConfigs := []aiv1beta1.RawNetConfig{}
+
+	// Select first the configs from the hosts, if defined
+	if agentConfig != nil {
+		for _, host := range agentConfig.Hosts {
+			// Skip worker hosts
+			if host.Role == "worker" {
+				continue
+			}
+			rawConfigs = append(rawConfigs, host.NetworkConfig.Raw)
+		}
+	}
+
+	// Then pick up the ones from nmStateConfigs, if defined
+	for _, nmStateConfig := range nmStateConfigs {
+		rawConfigs = append(rawConfigs, nmStateConfig.Spec.NetConfig.Raw)
+	}
+
+	// Try to look for an eligible IP
+	for _, raw := range rawConfigs {
+		nodeZeroIP, err := getFirstIP(raw)
```

Comment:
Arguably we should search all valid hosts for an IPv4 address before going back and searching for an IPv6 address only if none of them have IPv4.
In practice it probably doesn't matter though.